rasterio/fill.py

```diff
@@ -49,6 +50,16 @@ def fillnodata(
     out : numpy ndarray
         The filled raster array.
     """
+    if mask is None and hasattr(image, 'mask'):  # pragma: no cover
+        mask = ~image.mask
+    if not dtypes.is_ndarray(mask):
```

Comment:
This should move to `_fill.pyx` so that basic input validation is done at the lowest level.

Reply:
@brendan-ward well, maybe... but validating and raising errors from the pure Python module (fill.py) has a big advantage (as I see it): the end user will be able to catch the error with pdb and do some debugging and introspection that is much harder to do with the extension module (_fill.pyx).

Reply:
I wasn't thinking about debugging, so that is a good catch.  But if we return a `ValueError` with an obvious message, the need for pdb to debug that specific case is minimal.

You could do validation in both places.  What I was getting at was that - although we don't officially recommend it - it is possible to call the underlying functions in the `.pyx` files.  We've seen tricky bugs crop up because we weren't validating inputs right before passing them along to GDAL...

We have other Cython level functions that validate and raise exceptions, so there is good precedent.  (but no idea about ease of debugging)